exonum/src/runtime/communication_channel.rs

```diff
@@ -0,0 +1,92 @@
+// Copyright 2019 The Exonum Team
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use crate::runtime::{
+    dispatcher::Dispatcher,
+    mailbox::{Action, AfterRequestCompleted, BlockchainMailbox},
+    CallInfo, ConfigChange, ExecutionContext, ExecutionError, InstanceId,
+};
+
+/// Marker trait for `Supervisor` entities. Implementors of that trait can get access
+/// to the extended interface of `CommunicationChannel`.
+#[doc(hidden)]
+pub trait SupervisorAccess {}
+
+/// Communication channel is an proxy entity for performing service calls from
+/// other service instances.
+#[derive(Debug)]
+pub struct CommunicationChannel<'a, T = ()> {
+    pub(crate) mailbox: &'a BlockchainMailbox,
+    dispatcher: &'a Dispatcher,
+    phantom: std::marker::PhantomData<T>,
+}
+
+impl<'a, T> CommunicationChannel<'a, T> {
+    /// Creates a new communication channel.
+    pub(crate) fn new(mailbox: &'a BlockchainMailbox, dispatcher: &'a Dispatcher) -> Self {
+        Self {
+            mailbox,
+            dispatcher,
+            phantom: std::marker::PhantomData,
+        }
+    }
+
+    /// Call the corresponding runtime method.
+    pub fn call(
```

Comment:
If there is `call`, shall we also add here `call_async` as a replacement for/extension to ApiSender#submit? It, however, will be rather different as it will produce _num_validators_ transactions when called unconditionally.

Reply:
>It, however, will be rather different as it will produce num_validators transactions when called unconditionally.

Yep, I think that's the core point. Async transactions currently can be emitted only by users, since they sign them with their personal keys. I think implementing of such a feature will require a really hard R&D work inbefore it can be implemented.

Reply:
The thing is currently they can be emitted using ApiSender and signed by the node service key (time service uses this, probably, some configuration service uses this), and the service developers using this [has to be aware of the consequences](https://exonum.com/doc/api/java-binding/0.8.0/com/exonum/binding/core/service/Node.html#submitTransaction\(com.exonum.binding.core.transaction.RawTransaction\)).

Reply:
Well, I (from my side) would call that an architectural flaw. We need to develop a safe&consistent interface for that type of actions.